src/useSelections/index.ts

```diff
@@ -0,0 +1,80 @@
+/* eslint-disable no-shadow */
+import { useState, useMemo } from 'react';
+
+export default function useSelections<T>(items: T[]) {
+  const [selected, setSelected] = useState<T[]>([]);
+
+  const { selectedSet, isSelected, select, unSelect, toggle } = useMemo(() => {
+    const selectedSet = new Set<T>(selected);
+
+    const isSelected = (item: T) => selectedSet.has(item);
+
+    const select = (item: T) => {
+      selectedSet.add(item);
+      return setSelected(Array.from(selectedSet));
+    };
+
+    const unSelect = (item: T) => {
+      selectedSet.delete(item);
+      return setSelected(Array.from(selectedSet));
+    };
+
+    const toggle = (item: T) => {
+      if (isSelected(item)) {
+        unSelect(item);
+      } else {
+        select(item);
+      }
+    };
+
+    return { selectedSet, isSelected, select, unSelect, toggle };
+  }, [selected]);
+
+  const {
+    selectAll,
+    unSelectAll,
+    noneSelected,
+    allSelected,
+    halfSelected,
+    toggleAll,
+  } = useMemo(() => {
+    const selectAll = () => {
+      items.forEach(o => {
+        selectedSet.add(o);
+      });
+      setSelected(Array.from(selectedSet));
+    };
+
+    const unSelectAll = () => {
+      items.forEach(o => {
+        selectedSet.delete(o);
+      });
+      setSelected(Array.from(selectedSet));
+    };
+
+    const noneSelected = items.every(o => !selectedSet.has(o));
+
+    const allSelected = items.every(o => selectedSet.has(o)) && !noneSelected;
+
+    const halfSelected = !noneSelected && !allSelected;
```

Comment:
改成叫 `partiallySelected` 可能更贴切一些。。？`halfSelected` 的意思感觉像是“正好选择了**一半**”

```suggestion
    const partiallySelected = !noneSelected && !allSelected;
```

Reply:
https://ant.design/components/checkbox-cn/#Checkbox
indeterminate 我觉得也可以。

Reply:
@brickspert  你觉得如何

Reply:
partiallySelected 可以的。全选/部分选/没选 名字也统一一点。
indeterminate 不是很统一。
我投 partiallySelected 一票